src/main/java/io/cryostat/recordings/RecordingCreationHelper.java

```diff
@@ -0,0 +1,194 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.openjdk.jmc.common.unit.IConstrainedMap;
+import org.openjdk.jmc.flightrecorder.configuration.events.EventOptionID;
+import org.openjdk.jmc.flightrecorder.configuration.recording.RecordingOptionsBuilder;
+import org.openjdk.jmc.rjmx.services.jfr.IEventTypeInfo;
+import org.openjdk.jmc.rjmx.services.jfr.IRecordingDescriptor;
+
+import io.cryostat.commands.internal.EventOptionsBuilder;
+import io.cryostat.core.net.JFRConnection;
+import io.cryostat.core.templates.TemplateType;
+import io.cryostat.messaging.notifications.NotificationFactory;
+import io.cryostat.net.ConnectionDescriptor;
+import io.cryostat.net.TargetConnectionManager;
+import io.cryostat.net.web.http.HttpMimeType;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class RecordingCreationHelper {
+
+    private static final String NOTIFICATION_CATEGORY = "RecordingCreated";
+
+    private static final Pattern TEMPLATE_PATTERN =
+            Pattern.compile("^template=([\\w]+)(?:,type=([\\w]+))?$");
+
+    private final TargetConnectionManager targetConnectionManager;
+    private final EventOptionsBuilder.Factory eventOptionsBuilderFactory;
+    private final NotificationFactory notificationFactory;
+
+    RecordingCreationHelper(
+            TargetConnectionManager targetConnectionManager,
+            EventOptionsBuilder.Factory eventOptionsBuilderFactory,
+            NotificationFactory notificationFactory) {
+        this.targetConnectionManager = targetConnectionManager;
+        this.eventOptionsBuilderFactory = eventOptionsBuilderFactory;
+        this.notificationFactory = notificationFactory;
+    }
+
+    public IRecordingDescriptor startRecording(
+            ConnectionDescriptor connectionDescriptor,
+            IConstrainedMap<String> recordingOptions,
+            String templateName,
+            TemplateType templateType)
+            throws Exception {
+        String recordingName = (String) recordingOptions.get(RecordingOptionsBuilder.KEY_NAME);
+        return targetConnectionManager.executeConnectedTask(
+                connectionDescriptor,
+                connection -> {
+                    if (getDescriptorByName(connection, recordingName).isPresent()) {
+                        throw new IllegalArgumentException(
+                                String.format(
+                                        "Recording with name \"%s\" already exists",
+                                        recordingName));
+                    }
+                    IRecordingDescriptor desc =
+                            connection
+                                    .getService()
+                                    .start(
+                                            recordingOptions,
+                                            enableEvents(connection, templateName, templateType));
+                    notificationFactory
```

Comment:
Since automatic rules isn't part of the web client yet, who receives these notifications?

Reply:
Nobody, really. You could observe them using `websocat` if you wanted to. Otherwise, the notification will flow through to the messaging server and then be essentially dropped as there are no connected clients to write to.

Reply:
Oh wait - this is a notification from the `RecordingCreationHelper`. So this is a notification with category `RecordingCreated`, which the web-client does have handling for - it's used to refresh the recordings list view. If you're connected to Cryostat via the web-client and are looking at the recordings list view for a target, and Cryostat creates a recording on that target due to some new Rule being added (at this point this probably means you just did that yourself manually with an API request), you'll see the web-client refresh itself to see the new recording appear.

Reply:
Just to clarify, what this means is that from any notifications channel client's perspective, any `RecordingCreated` notification is indistinguishable. You could connect using the web-client and observe that it receives these notifications:

1. when the same web-client instance performs an API request to create a recording
2. when a _different_ web-client instance connected to the same backend performs an API request to create a recording
3. when the Operator performs an API request to create a recording (assuming this Cryostat backend is deployed by an Operator)
4. when some user, yourself or otherwise, uses curl/httpie to perform an API request to create a recording
5. when Cryostat itself decides to create a recording, because of automatic rule processing or hypothetically any other reason

etc. This is generally how the whole notifications channel system works - Cryostat emits messages when it does certain things, regardless of the reason that thing was done, and any clients who happen to be listening at that time are pushed a notification to inform them of the change.